### PR TITLE
LTD-4652: Manage concurrent SMTP connections using a lock

### PR DESCRIPTION
**Pull request by saruniitr**
## Change description

We use SMTP to send emails. As we process messages we also have a requirement to send emails from multiple tasks, because of this we may open multiple SMTP connections concurrently. This results in error if the number of concurrent connections exceed maximum allowed value.

To manage this reliably we are restricting access to this shared resource using a lock. This is achieved by deferring all email sending functionality to a dedicated task.
Before sending email it first tries to acquire a lock.
  - If there are no active connections then it acquires lock and sends email.
     In some cases we need to update state which is handled in subtasks linked to the parent task.
  - If there is active connection (lock acquisition fails) then it raises an exception to trigger an automatic retry.
     If all retries fail then manual intervention may be required (unlikely)

`Co-authored-by: Le H Ngo <hgle.ngo@gmail.com>`

[LTD-4652](https://uktrade.atlassian.net/browse/LTD-4652)

[LTD-4652]: https://uktrade.atlassian.net/browse/LTD-4652?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: mail/libraries/routing_controller.py [hunk 1]
```diff
@@ -175,6 +175,8 @@ def send(email_message_dto: EmailMessageDto):
 
 
 def _collect_and_send(mail: Mail):
+    from mail.celery_tasks import send_email_task
+
```
Comment (saruniitr) [inline mail/libraries/routing_controller.py:179]: unavoidable due to circular dependencies

File: mail/tests/test_send_lite_licence_updates_task.py [hunk 2]
```diff
@@ -1,49 +1,84 @@
...
-            (ReceptionStatusEnum.REPLY_PENDING, 1, 1, 0),
-            (ReceptionStatusEnum.REPLY_RECEIVED, 1, 1, 0),
-            (ReceptionStatusEnum.REPLY_SENT, 2, 1, 1),
+            ([True],),
+            ([False, True],),
+        ]
+    )
+    @mock.patch("mail.celery_tasks.cache")
+    @mock.patch("mail.celery_tasks.smtp_send")
+    def test_send_licence_details_success(self, lock_acquired, mock_smtp_send, mock_cache):
```
Comment (saruniitr) [inline mail/tests/test_send_lite_licence_updates_task.py:27]: Ideally this test should be in `test_celery_tasks.py` but existing tests for this task are in a separate file so added it here.

**Comment by saruniitr** [comment]
> I am not sure since ur are not injecting straight into `send` its still being used by `parse_and_reply_emails` and there is management command which is using but I guess this not matter as much, would not go directly into the send better as I did?

You mean call `send()` directly instead of below?
```
            if email_type == EmailType.LITE_LICENCE_DETAILS or email_type == EmailType.SPIRE_LICENCE_DETAILS:
                message = build_email_message(message_dto)
                smtp_send(message)

            elif email_type == EmailType.LICENCE_REJECTED:
                message = build_licence_rejected_email_message(message_dto)
                smtp_send(message)
```

Yes, we can call `send()` directly in the first `if` condition but to keep it consistent with the second `if` condition I am calling individual functions.

You can ignore `parse_and_reply_emails()` which part of `mock_hmrc` as it is incomplete. I think we can remove that completely.

Yes, `send()` is also called in a management command `resend_email()`, I considered changing that to a task as well but that is not required as we only use it manually in special circumstances.

File: mail/celery_tasks.py [hunk 3]
```diff
@@ -75,38 +80,132 @@ def _log_error(message, lite_usage_data_id):
...
+            logger.exception("Another SMTP connection is active, will be retried after backing off")
+            raise SMTPConnectionBusy()
+
+        logger.info("Lock acquired, proceeding to send email")
+
+        # de-serialize back to original type
+        message_dto = EmailMessageDto(*message_dto)
+
+        message = None
+        if email_type == EmailType.LITE_LICENCE_DETAILS or email_type == EmailType.SPIRE_LICENCE_DETAILS:
```
Comment (saruniitr) [inline mail/celery_tasks.py:176]: this type is introduced to identify which task is sending this email.
Each task has a different set of arguments and follow-up actions after sending an email.
It may be possible to identify based on other arguments but made it explicit.

Eg the task that send licences rejected notification email only needs to send `message_dto`. It can send `mail_id` as `None` and we can use that to identify but that will introduce additional if conditions.

The other three instances where this is called is when SPIRE sends new licence details, reply back and usage data back - in all these cases we need `mail_id, dto` and update them as follow up tasks. 

The only task that provides `licence_payload_ids` is the one that sends LITE licence details hence it is assigned a default value so other 3 tasks can ignore this.

Comment (saruniitr) [reply mail/celery_tasks.py:176]: We can try to pass the final message to be sent as argument so this task would simply call `smtp_send(message)` but we still need `Mail, Dto` arguments to perform follow-up actions.

File: mail/tests/test_resend_email.py [hunk 4]
```diff
@@ -62,19 +69,26 @@ def test_resend_licence_data_mail_to_hmrc(
         self.assertEqual(mail.id, pending_mail.id)
         self.assertEqual(mail.status, ReceptionStatusEnum.REPLY_PENDING)
         self.assertEqual(mail.extract_type, ExtractTypeEnum.LICENCE_DATA)
-        self.assertEqual(send_mail.call_count, 2)
+        self.assertEqual(mock_send.call_count, 1)
 
+    @override_settings(SEND_REJECTED_EMAIL=False)
```
Comment (saruniitr) [inline mail/tests/test_resend_email.py:74]: this just interferes with the call count value because the reply message content has rejections so disabled this as it is not related to this particular test.

**Comment by hnryjmes** [comment]
@saruniitr has this been tested on a dev / demo environment?

File: mail/celery_tasks.py [hunk 5]
```diff
@@ -75,38 +80,132 @@ def _log_error(message, lite_usage_data_id):
...
+        licence_payload_ids = kwargs.get("licence_payload_ids", [])
+        if email_type == EmailType.LITE_LICENCE_DETAILS and licence_payload_ids:
+            # Mark payloads as processed
+            licence_payloads = LicencePayload.objects.filter(id__in=licence_payload_ids)
+            licence_payloads.update(is_processed=True)
+
+            references = licence_payloads.values_list("reference", flat=True)
+            logger.info("Licence payloads of references %s marked as processed", references)
+
+        return super().on_success(retval, task_id, args, kwargs)
```
Comment (currycoder) [inline mail/celery_tasks.py:125]: We can do callbacks with celery, which might make this a bit cleaner.  https://docs.celeryq.dev/en/stable/userguide/calling.html#linking-callbacks-errbacks
Gurdeep implemented something similar here; https://github.com/uktrade/lite-api/pull/1811/files#diff-ba75c0a31a6b6ed70ec3adffebfd6843645ab9dde64b35fe1c49092919e2857dR22

Comment (currycoder) [reply mail/celery_tasks.py:125]: Benefit of this is that we can have the "success operations" defined by the caller to `send_email_task` rather than baked in to here with a switch statement

File: mail/celery_tasks.py [hunk 6]
```diff
@@ -75,38 +80,132 @@ def _log_error(message, lite_usage_data_id):
...
+            message = build_licence_rejected_email_message(message_dto)
+
+        try:
+            if message:
+                smtp_send(message)
+        except SMTPException:
+            logger.exception("An unexpected error occurred when sending email -> %s")
+            raise
+
+        logger.info("Email sent successfully")
```
Comment (hnryjmes) [inline mail/celery_tasks.py:188]: is `message` is None it looks like you will still get to the `logger.info` statement 

wouldn't you want the success message to be inside the try block and the failure message to be the fall through option?

Comment (saruniitr) [reply mail/celery_tasks.py:188]: I will move this to within the if condition.

File: mail/celery_tasks.py (hunk 3 above)
Comment (currycoder) [reply mail/celery_tasks.py:176]: I realise you're heading this comment off, but this pattern where we need the guts of our email sending to do conditional things to build/finalise an email is a bit of a shame.  Did you try doing the custom serialization?  What made it difficult?

**Review by hnryjmes — APPROVED** [review]

File: mail/celery_tasks.py (hunk 3 above)
Comment (saruniitr) [reply mail/celery_tasks.py:176]: I have not tried that as using the type that can be avoided.
We need custom serialization for two models but with linking the tasks it may also be not necessary.

Comment (saruniitr) [reply mail/celery_tasks.py]: @currycoder updated to use task specific serializer.

File: mail/celery_tasks.py [hunk 7]
```diff
@@ -75,38 +80,106 @@ def _log_error(message, lite_usage_data_id):
...
 
 
-# Notify Users of Rejected Mail
+class SMTPConnectionBusy(SMTPException):
+    pass
+
+
+@contextmanager
+def cache_lock(lock_id):
+    timeout_at = time.monotonic() + LOCK_EXPIRE - 3
```
Comment (currycoder) [inline mail/celery_tasks.py:94]: what's the `- 3` bit doing?

**Review by currycoder — APPROVED** [review]
Approved, pending code coverage - just one question for my understanding

File: mail/celery_tasks.py (hunk 7 above)
Comment (saruniitr) [reply mail/celery_tasks.py:94]: I am also not sure, this is taken from the example in Celery documentation.
